### PR TITLE
[RFC] vim-patch:7.4.{831,832,845,903}

### PR DESCRIPTION
**Pull request by watiko**
```
vim-patch:7.4.831

Problem:    When expanding  on the command line and encountering an
            error, the command is executed anyway.
Solution:   Bail out when an error is detected.
```

https://github.com/vim/vim/commit/3f188935ec4db5117c4a64cc3f71219175624745

```
vim-patch:7.4.832

Problem:    $HOME in  is expanded too early.
Solution:   Skip over  when expanding environment names.
```

https://github.com/vim/vim/commit/be83b73ddb2ee8297037166d243f72e3423a3ce3

```
vim-patch:7.4.845

Problem:    Compiler warning for possible loss of data.
Solution:   Add a type cast. (Erich Ritz)
```

https://github.com/vim/vim/commit/5df1ed2de3fa9dcace996b9a0a4c9b3cea79cf1e

```
vim-patch:7.4.903

Problem:    MS-Windows: When 'encoding' differs from the current code page,
            expandinig wildcards may cause illegal memory access.
Solution:   Allocate a longer buffer. (Ken Takata)
```

https://github.com/vim/vim/commit/7314efd87d8c4095229bdc2867a553c36c064918

### CODE REVIEW

File: src/nvim/os/env.c [hunk 1]
```diff
@@ -262,8 +262,27 @@ void expand_env_esc(char_u *srcp, char_u *dst, int dstlen, bool esc, bool one,
...
+    if (src[0] == '`' && src[1] == '=') {
+      size_t len;
+
+      var = src;
+      src += 2;
+      (void)skip_expr(&src);
+      if (*src == '`') {
+        src++;
+      }
+      len = (size_t)src - (size_t)var;
```
Comment (oni-link) [inline src/nvim/os/env.c]: Declaration of `len` can be here.

File: src/nvim/os/env.c [hunk 2]
```diff
@@ -262,8 +262,27 @@ void expand_env_esc(char_u *srcp, char_u *dst, int dstlen, bool esc, bool one,
...
+      if (len > (size_t)dstlen) {
+        len = (size_t)dstlen;
```
Comment (oni-link) [inline src/nvim/os/env.c]: After the outer `while` loop `*dst = NUL` is used, so we cannot use up all `dstlen` bytes here.

File: src/nvim/os/env.c [hunk 3]
```diff
@@ -262,8 +262,27 @@ void expand_env_esc(char_u *srcp, char_u *dst, int dstlen, bool esc, bool one,
...
+      }
+      xstrlcpy((char *)dst, (char *)var, len);
```
Comment (oni-link) [inline src/nvim/os/env.c]: If we have to save up space for the `NUL` byte after the loop, we can use `memcpy()` here.

**Comment by watiko** [comment]
https://github.com/watiko/neovim/commit/209d844f021e87df0d16586e3e44e0465a1b6ac1

@oni-link Done. Many thanks.

**Pull request merged**